docs/design/api-versions.md

```diff
@@ -32,8 +40,169 @@ For each API spec, generated `AssignPropertiesTo*()` method  will read the value
+
+To keep the reconciler a simple as possible, we'll handle this with generated code hidden behind a simple interface:
+
+``` go
+type GeneratedResource interface {
+    // GetSpec returns the spec of the resource ready for 
+    // serialization to ARM
+    GetSpec(scheme runtime.Scheme) (genruntime.ARMTransformer, error)
+
+    // NewStatus returns a new, empty, status instance ready for
+    // population from ARM
+    NewStatus(scheme runtime.Scheme) (genruntime.FromARMConverter, error)
+
+    // SetStatus updates the current resource with the status provided
+    // The instance passed here should always be the one returned by
+    // NewStatus()
+    SetStatus(status genruntime.FromARMConverter) error
+}
+```
+
+The implementation strategy used for API types and Storage types will differ, as detailed below. Conversion between versions of a resource requires access to a `runtime.Scheme` instance, which is why each method has one as a parameter.
+
+### Implementation for API versions
+
+For all API versions of resources, the generated implementation of this interface will work directly with the `Spec` and `Status` of that API type.
+
+``` go
+// GetSpec returns the Spec of this resource
+func (p *Person) GetSpec(_ runtime.Scheme) (genruntime.ARMTransformer, error) {
+    return &p.Spec, nil
+}
+
+// NewStatus returns a new blank status, ready to be populated
+func (p *Person) NewStatus(_ runtime.Scheme) (genruntime.FromARMConverter, error) {
+    return &Person_Status{}, nil
+}
+
+// SetStatus accepts a status and updates this resource
+// Only a status of the expected type is handled; anything else will error
+func (p *Person) SetStatus(status interface{}) error {
+    s, ok := status.(*Person_Status);
+    if !ok {
+        return errors.Errorf(
+            "expected Status of type %T, but received %T", p.Status, status)
+    } 
+
+    p.Status = *s
+    return nil
+}
+```
+
+### Implementation for storage versions
+
+For the hub Storage version of resources, the generated implementation will transparently handle conversion between versions as dictated by the `OriginalVersion()` function described earlier.
+
+We pivot to the original version by using the `OriginalGVK()` function to obtain a new, empty resource from the schema.
+
+``` go
+// GetSpec returns the spec of the current type using 
+// the original API version from when the CR was created
+func (p *Person) GetSpec(scheme runtime.Scheme) (genruntime.ARMTransformer, error) {
+    gvk := p.OriginalGVK()
+    
+    // Use gvk to create an empty resource of the expected version 
+    originalObj, err := scheme.New(gvk)
+    if err != nil {
+        return errors.Wrapf(
+            err, "unable to create requested version %s of resource", gvk)
+    }
+    resource, ok := originalObj.(genruntime.GeneratedResource)
+    if !ok {
+        return errors.Wrapf(
+            err, "unable to convert %T to genruntime.GeneratedResource", originalObj)
```

Comment:
minor: Don't want to wrap `err` here because the issue was `!ok` not `err != nil`

Reply:
Fixed.